# Exploring Ebay Car Sales Data

What is given? Given a dataset of used cars from eBay Kleinanzeigen (a classifieds section of the German eBay website), which was originally uploaded to Kaggle, but later modified by Dataquest team for studying purposes. It was sampled to only 50.000 data points and carefuly dirtied so that we, aspiring data analytics, could check if the common saying about data scientists spending 80% of their time to clean the data is True or False. 

Therefore, the objective of this project is to clean the data, analyze it and see what insights there're behind the data.

So, let's do it!



## Import the libraries, read in the file, initial exploration

They say that the NumPy and pandas libraries are essential for working with data. So, let´s listen to them.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Here it is. A dataset with 50000 entries with all the data structured in 20 columns. And as promised, a few issues can be spotted by just looking and no coding:
 - column labels which are difficult to read 
 - missing values in `vehicleType` , `gearbox`, `model`, `fuelType` and `notReparedDamage` columns (for neigther of them the null values exceed 20% of all the values)
 - datetimelike data stored as string (see `dateCrawled`, `dateCreated` and `lastSeen` columns)
 - numeric data `odometer` and `price` columns overwhelemed by special characters and thus stored as string instead of integer
 - really weird description table for `nrOfPictures` column (it´s better take a better look later)
 - some columns are not going to give us any insight as almost all their values are the same (check `seller` and `offerType` columns)
 - German words, many German words used for values in various columns (one would never expect it from a data scrapped from a German website, eh?😉)

## Cleaning Data. Part I

### Updating Column Labels

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

The job to be done here is to make the column labels more readable and more accurate. Camelcase will be transformed to snakecase, and some optimization will be implemented as well.

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [7]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

### Too Many Strings... Or...Changing Dtypes of Relevant Columns

In [8]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
autos['ad_created'] = pd.to_datetime(autos['ad_created']) #the timestamp is lost during the convertion
autos['last_seen'] = pd.to_datetime(autos['last_seen'])

In [9]:
#removing all the non-numeric characters and converting to integers
autos['odometer'] = (autos['odometer']
                            .str.replace(',', '')
                            .str.replace('km', '')
                            .astype(int)
                    )

autos['price'] = (autos['price']
                          .str.replace('$', '')
                          .str.replace(',', '')
                          .astype(int)
                 )

#adding units of measure to the 'odometer' and 'price' column names
autos.rename({'odometer':'odometer_km', 'price': 'price_$'}, axis = 1, inplace = True)

### `'seller'`, `'offer_type'` and `'num_photos'`, Do we really need you?

In [10]:
print("Unique values in 'seller' column:", autos['seller'].unique())
print("Unique values in 'offer_type' column:", autos['offer_type'].unique())

Unique values in 'seller' column: ['privat' 'gewerblich']
Unique values in 'offer_type' column: ['Angebot' 'Gesuch']


We knew it already, that these columns were not going to provide us with many information. And now we know that the little information they have is all in German. No worries, Google.Translate works better every day. The `'seller'` column  is all about if a listing is a private one or a commercial one (private in almost all the cases), and the `'offer_type'` - if it´s a 'bid' or 'application' offer. Curious but not informative. So, it´s better drop these columns and save some screen space. 

In [11]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

The `'num_photos'` column is quite boring, it´s got only 0 for each listing. Have you ever considered to buy a car without seeing any its photo? Probably there´s been some error while scraping data. So, dropping this column as well.

In [12]:
autos = autos.drop(['seller', 'offer_type', 'num_photos'], axis=1)

### Google.Translate, we need you again

It´s a turn of Google.Translate again. Let´s just replace the German words in `'vehicle_type'`, `'gearbox'`, `'fuel_type'`, `'model'`, `'unrepaired_damage'`and `'brand'`columns to its English translations and we´ll be done with cleaning data for now.

In [13]:
translations = {"kleinwagen": "super mini", "kombi":"station wagon", "cabrio": "convertible", "limousine": "sedan", "andere": "other", "manuell": "manual", "automatik": 
            "automatic", "benzin":"gas", "elektro": "electric", "sonstige_autos": "other", "nein": "no", "ja": "yes"}

columns_to_change = ['vehicle_type', 'gearbox', 'fuel_type', 'model', 'unrepaired_damage', 'brand']

autos[columns_to_change] = autos[columns_to_change].replace(translations)

## Detecting Outliers and Other Oddities

### Exploring `'odometer_km'` column

In [14]:
print("Number of unique values in 'odometer_km' column:", 
      autos['odometer_km'].unique().shape[0])
print('\n')

print("Values distribution in 'odometer_km' column:", '\n', 
      autos['odometer_km'].describe())
print('\n')

print("Counts of unique values for 'odometer_km' column, in percentages:", '\n', 
      autos['odometer_km'].value_counts(normalize=True).sort_index(ascending=False) * 100)

Number of unique values in 'odometer_km' column: 13


Values distribution in 'odometer_km' column: 
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


Counts of unique values for 'odometer_km' column, in percentages: 
 150000    64.848
125000    10.340
100000     4.338
90000      3.514
80000      2.872
70000      2.460
60000      2.328
50000      2.054
40000      1.638
30000      1.578
20000      1.568
10000      0.528
5000       1.934
Name: odometer_km, dtype: float64


What do these numbers tell us? 

First of all, the values of mileage are rounded off. Probably one is not allowed to fill the value manually but is offered to choose from some pre-set options of the value.

Secondly, there are barely cars that has just left the dealer´s shop. Almost 80% of the cars has got 100.000 or more kilometers done. Nothing surprising so far - we are exploring used car sales data. What may seem kind of unusual is that the cars with 150.000 or more mileage account for almost 65% of all the listing. Before discarding them as outliers let's explore the dataset more in depth - there might be some explanation to this fact.     

### Exploring `'price_$'` column 

In [15]:
print("Number of unique values in 'price_$' column:", autos['price_$'].unique().shape[0])
print('\n')

print("Values distribution in 'price_$' column:", '\n', autos['price_$'].describe())
print('\n')

print("Counts of highest values for 'price_$' column:", '\n', 
      autos['price_$'].value_counts().sort_index(ascending=False).head(20))
print('\n')

print("Counts of lowest values for 'price_$' column:", '\n',
      autos['price_$'].value_counts().sort_index().head(20))

Number of unique values in 'price_$' column: 2357


Values distribution in 'price_$' column: 
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


Counts of highest values for 'price_$' column: 
 99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_$, dtype: int64


Counts of lowest values for 'price_$' column: 
 0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64


What have we gotten here?
Firstly, most of the values seem to be rounded as well as in the `'odometer_km'` column. But rather than having pre-set options, I would opt for the people tending to round off the price (considering the 2357 unique values).

Secondly, there are 1421 cars with \\$0 price. We don´t know if the owners of these cars try to get rid off their very old and probably damaged cars and save on the disposal tax or there's some error in these entries. Anyway, \$0 price cars account for less than 2% of all the listings and we can remove these rows without affecting too much the results of the analysis. 

Thirdly, there´s quite a number of cars which are sold for less than \\$35. It might be just fine if we remember that eBay is an online auction. So we´ll keep the items with at least \$1 opening bid.

When everything is clear with the cheap cars, let´s move on to the most expensive ones. When examining records above \\$350,000, several of these entries appear to be illegitimate entries (e.g., 1234566, 12345678, 1111111, 99999999). In addition, it is unlikely that these prices represent the value of the vehicles given that this is a used vehicle classified site. Therefore, anything below 1.00 and above \$350,000 will be removed as these could have been errors or illegitimate entries. 

In [16]:
#removing rows with the price less than $1 and more than $350,000
autos = autos[autos['price_$'].between(1,350000)]

autos['price_$'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

## Dealing with date columns

### 'date_crawled', 'ad_created', 'last_seen'

In [17]:
#exploring the 'date_crawled' column:
print("Counts of unique values for 'date_crawled' column, in percentages:", '\n', 
      autos['date_crawled'].dt.date.value_counts(normalize=True, dropna=False).sort_index() * 100)

Counts of unique values for 'date_crawled' column, in percentages: 
 2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64


No anomalities with the `'date_crawled'` column (something one would expect from the column created by the crawler). All the data was scrapped between March 05, 2016 and April 07, 2016. We´ll remember this dates for further analysis.

In [18]:
#exploring the 'ad_created' column:
print("Counts of unique values for 'ad_created' column, in percentages:", '\n', 
      autos['ad_created'].dt.date.value_counts(normalize=True, dropna=False).sort_index() * 100)

Counts of unique values for 'ad_created' column, in percentages: 
 2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: ad_created, Length: 76, dtype: float64


So far, no huge anomalities in the 'ad_created' column as well. Just one a little bit weird thing. For a period between June 11, 2015 and April 07, 2016 the 48565 ads were created on only 76 different days. 

In [19]:
#first 15 entries created
(autos['ad_created'].dt.date
     .value_counts(normalize=True, dropna=False)
     .sort_index()
     .head(15)* 100)

2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
2015-12-30    0.002059
2016-01-03    0.002059
2016-01-07    0.002059
2016-01-10    0.004118
2016-01-13    0.002059
2016-01-14    0.002059
2016-01-16    0.002059
2016-01-22    0.002059
2016-01-27    0.006177
2016-01-29    0.002059
Name: ad_created, dtype: float64

The mistery is solved. There´re only 6 ads created in 2015, which definitely can´t be legitimate. It should be some specifics of crawled data. To confirm that, the documentation to the original data set should be revised.

In [20]:
#exploring the 'last_seen' column:

print("Counts of unique values for 'last_seen' column, in percentages:", '\n', 
      autos['last_seen'].dt.date.value_counts(normalize=True, dropna=False).sort_index() * 100)

Counts of unique values for 'last_seen' column, in percentages: 
 2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.


### What does the registration year tell us?

In [21]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The general info about the registration year gives us 2 main insights. There are cars in our dataset that had been registered some 8 centuries before cars were invented (1000 is the min registration year) and there are cars from the future as the maximum registration year is stated as 9999. We can assume that among all the cars there might be some retro cars. So let´s take 1900 as an intial year and the final year will be the year when the data was scrawled, 2016.  

In [22]:
incorrect_years_percentage = (~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0] * 100
print(incorrect_years_percentage)

3.8793369710697


There´re less than 4% of listings with incorrect registration year. It´s safe to remove them.

In [23]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [24]:
print("Counts of unique values for 'registration_year' column, in percentages:", '\n', 
      autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head(30)* 100)

Counts of unique values for 'registration_year' column, in percentages: 
 2016    2.613483
2015    0.839742
2014    1.420278
2013    1.720186
2012    2.806281
2011    3.476789
2010    3.403954
2009    4.466485
2008    4.744971
2007    4.877788
2006    5.719672
2005    6.289497
2004    5.790364
2003    5.781796
2002    5.325507
2001    5.646837
2000    6.760781
1999    6.205951
1998    5.062017
1997    4.179431
1996    2.941239
1995    2.628478
1994    1.347443
1993    0.910435
1992    0.792614
1991    0.726206
1990    0.743343
1989    0.372743
1988    0.289197
1987    0.154238
Name: registration_year, dtype: float64


It appears that most of the cars for sale were registered starting from the early 90s. 

## Done with cleaning, let´s the analysis begin

### The Brands and The Prices

In [25]:
print("Counts of unique values for 'brand' column, in percentages:", '\n',
      autos['brand'].value_counts(normalize=True) * 100)

Counts of unique values for 'brand' column, in percentages: 
 volkswagen       21.126368
bmw              11.004477
opel             10.758124
mercedes_benz     9.646323
audi              8.656627
ford              6.989996
renault           4.714980
peugeot           2.984083
fiat              2.564212
seat              1.827296
skoda             1.640925
nissan            1.527388
mazda             1.518819
smart             1.415994
citroen           1.400998
toyota            1.270324
hyundai           1.002549
other             0.981127
volvo             0.914719
mini              0.876159
mitsubishi        0.822604
honda             0.784045
kia               0.706926
alfa_romeo        0.664082
porsche           0.612669
suzuki            0.593389
chevrolet         0.569825
chrysler          0.351321
dacia             0.263490
daihatsu          0.250637
jeep              0.227073
subaru            0.214220
land_rover        0.209936
saab              0.164949
jaguar            0.

What a surprise! We´re analysing a data set of used cars listed on the German section of eBay and the top 5 most represented brands come from Germany accounting for almost 60% of all the listing in total. Germans are loyal to German cars, confirmed. 

More than half of the brands are represented quite poorly and account for less than 1% of all the listings. We are not going to consider them for the further analysis.

In [26]:
#selecting the most common brands
brand_count = autos['brand'].value_counts(normalize=True) * 100
common_brand = brand_count[brand_count>1].index
print(common_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai'],
      dtype='object')


Let´s check now whar are the average prices for each brand.

In [27]:
#calculating the mean price by brand
def mean_value_by_brand(column, dataframe):
    """ calculate mean value of a parameter (price, mileage, etc.) for each brand out of defined list most common brand"""
    mean_value_by_brand = {}
    for brand in common_brand:
        mean_value = dataframe[dataframe['brand'] == brand][column].mean()
        mean_value_by_brand[brand] = mean_value
    return pd.Series(mean_value_by_brand).astype(int).sort_values(ascending=False)


mean_prices = mean_value_by_brand(column = 'price_$', dataframe = autos)
mean_prices

audi             9336
mercedes_benz    8628
bmw              8332
skoda            6368
volkswagen       5402
hyundai          5365
toyota           5167
nissan           4743
seat             4397
mazda            4112
citroen          3779
ford             3749
smart            3580
peugeot          3094
opel             2975
fiat             2813
renault          2474
dtype: int32

The most expensive cars are expectedly Audi, Mercedes and BMW cars. The price of the most represented Wolkswagen cars is in the middle range. This may be a factor for its popularity.

 ### The Brands, The Prices and The Mileage

The same way we are going to analyze the mean mileage per brand and then aggregate all together to a dataframe for comparing.

In [28]:
#calculating the mean mileage by brand

mean_mileages = mean_value_by_brand('odometer_km', autos)
mean_mileages


bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
renault          128071
peugeot          127153
mazda            124464
ford             124266
seat             121131
citroen          119694
nissan           118330
fiat             117121
toyota           115944
skoda            110848
hyundai          106442
smart             99326
dtype: int32

In [29]:
#aggregating mean prices and mileages to one dataframe
brand_info = pd.DataFrame(mean_prices.astype(int), columns=['mean_price'])
brand_info['mean_mileage'] = mean_mileages.astype(int)
brand_info

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
skoda                6368        110848
volkswagen           5402        128707
hyundai              5365        106442
toyota               5167        115944
nissan               4743        118330
seat                 4397        121131
mazda                4112        124464
citroen              3779        119694
ford                 3749        124266
smart                3580         99326
peugeot              3094        127153
opel                 2975        129310
fiat                 2813        117121
renault              2474        128071

First of all, there´s no much difference in the mean mileage by brand. And there´s a slight tendency that more expensive cars has got more mileage. What can this tendency tell us? First, the owners of BMW, Mercedes, Audi tend to give more ride to their cars before selling them. Second, when selling, they consider the car to be in a condition  good enought to ask quite a higher price for it. The popular Wolkswagen cars have got quite a decent combination of price and mileage, which add one more brick to its popularity factors.    

### Look, but what if a car is damaged?

True to say, there´re almost 4540 cars which has got some unrepaired damaged mentioned by its owners. We can check if these listings are somewhat different in terms of price and mileage.

In [30]:
autos['unrepaired_damage'].value_counts(dropna=False)

no     33834
NaN     8307
yes     4540
Name: unrepaired_damage, dtype: int64

In [31]:
#getting the mean values by unrepaired_damage status and brand
brand_info_damages = autos.groupby(['unrepaired_damage', 'brand'])[['price_$', 'odometer_km']].mean().astype(int)

#unstacking the outer level
brand_info_damages = brand_info_damages.unstack(level=0)

#retrieving only the defined most common brands
brand_info_damages = brand_info_damages.loc[common_brand]

#renaming columns
columns = pd.MultiIndex.from_product([['mean_price', 'mean_mileage'], ['no', 'yes']], names=['', 'unrepaired_damage'])
brand_info_damages.columns = columns

#sorting by mean price for 'no unrepaired damage' marked cars
brand_info_damages = brand_info_damages.sort_values(by=('mean_price', 'no'), ascending=False)
brand_info_damages

mean_price       mean_mileage        
unrepaired_damage         no   yes           no     yes
audi                   10914  3324       125505  143198
mercedes_benz           9798  3921       129389  139883
bmw                     9437  3512       131115  141215
skoda                   7053  3565       108259  117835
volkswagen              6469  2179       125718  137718
hyundai                 6194  2417       101196  127543
nissan                  5765  1947       113990  133118
toyota                  5569  3327       114197  123658
seat                    5183  1741       117793  135406
mazda                   5110  1418       121402  132134
ford                    4660  1375       121343  134710
citroen                 4289  1944       116339  132573
smart                   3927  1445        95468  122000
peugeot                 3674  1372       123156  139331
opel                    3660  1367       127022  134771
fiat                    3446  1146       112118  130000
renault                 3103  1145       124783  134751

The general trend is that the cars which are claimed to have some damage to repair are sold quite cheaper and have got somewhat more mileage. 

In [32]:
brand_info_damages[('mean_price_deviation', '')] = (brand_info_damages[('mean_price', 'no')] - brand_info_damages[('mean_price', 'yes')])/ brand_info_damages[('mean_price', 'no')] * 100
brand_info_damages[('mean_mileage_deviation', '')] = (brand_info_damages[('mean_mileage', 'yes')] - brand_info_damages[('mean_mileage', 'no')])/ brand_info_damages[('mean_mileage', 'yes')] * 100
brand_info_damages

mean_price       mean_mileage         mean_price_deviation  \
unrepaired_damage         no   yes           no     yes                        
audi                   10914  3324       125505  143198            69.543705   
mercedes_benz           9798  3921       129389  139883            59.981629   
bmw                     9437  3512       131115  141215            62.784783   
skoda                   7053  3565       108259  117835            49.454133   
volkswagen              6469  2179       125718  137718            66.316278   
hyundai                 6194  2417       101196  127543            60.978366   
nissan                  5765  1947       113990  133118            66.227233   
toyota                  5569  3327       114197  123658            40.258574   
seat                    5183  1741       117793  135406            66.409415   
mazda                   5110  1418       121402  132134            72.250489   
ford                    4660  1375       121343  134710            70.493562   
citroen                 4289  1944       116339  132573            54.674749   
smart                   3927  1445        95468  122000            63.203463   
peugeot                 3674  1372       123156  139331            62.656505   
opel                    3660  1367       127022  134771            62.650273   
fiat                    3446  1146       112118  130000            66.744051   
renault                 3103  1145       124783  134751            63.100226   

                  mean_mileage_deviation  
unrepaired_damage                         
audi                           12.355619  
mercedes_benz                   7.501984  
bmw                             7.152215  
skoda                           8.126618  
volkswagen                      8.713458  
hyundai                        20.657347  
nissan                         14.369206  
toyota                          7.650940  
seat                           13.007548  
mazda                           8.122058  
ford                            9.922797  
citroen                        12.245329  
smart                          21.747541  
peugeot                        11.609046  
opel                            5.749753  
fiat                           13.755385  
renault                         7.397348

One might get some ideas from this table on deciding when is the prefect moment to put one´s car for a sale and get the best price for it before an important failure of some part which is too costly to repaire and can drop the price by 60% or even more. 

### A bit more practicing with `.groupby()` method

Working on the previous problem I´ve discovered the `.groupby()` method and got fascinated about it. So while practicing a little bit more with it, let´s see whether average prices follows any patterns if we split all the mileages into several groups and let´s check as well the most popular models for each brand.

In [33]:
#splitting the mileage into groups
autos['mileage_group'] = pd.cut(autos['odometer_km'], bins=5)

#defining the size of each group
mileage_groups = autos.groupby('mileage_group').size()

#defining the mean price for each group
mileage_groups_price = autos.groupby('mileage_group')['price_$'].mean()

#combining the size and the mean price into one dataframe
mileage_group = pd.DataFrame(mileage_groups, columns=['counts'])
mileage_group['mean_price'] = mileage_groups_price.astype(int)
mileage_group


counts  mean_price
mileage_group                           
(4855.0, 34000.0]       2528       15122
(34000.0, 63000.0]      2918       13721
(63000.0, 92000.0]      4235        9563
(92000.0, 121000.0]     2058        8132
(121000.0, 150000.0]   34942        4107

Predictably the less kilometers done, the higher is the price. It´s interesting that the mean price goes down gradually for the first four groups and then drops by  50% for the cars with more than 121000 km done. 

In [34]:
popular_models = autos.groupby(['brand', 'model']).size().sort_values(ascending=False)
popular_models = popular_models.loc[common_brand]
popular_models.head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
dtype: int64

As for the most popular models, in the top 10 we can find 4 models of Volkswagen, 2 models of Opel, BMW and Mercedes-Benz and 1 model of Audi. The leader is Volkswagen Golf.

Below we can check the top 3 models for each common brand which we defined earlier.

In [35]:
#sorting the models within each brand 
popular_models_by_brand = (popular_models.reset_index()
        .sort_values(['brand', 0], ascending=[1,0])
        .set_index(['brand', 'model']))

#getting top 3 models for each brand
popular_models_by_brand.groupby(level=0).head(3)

0
brand         model         
audi          a4        1231
              a3         825
              a6         797
bmw           3er       2615
              5er       1132
              1er        521
citroen       other      219
              berlingo   101
              c5          72
fiat          punto      415
              other      229
              500        120
ford          focus      762
              fiesta     722
              mondeo     479
hyundai       i_reihe    174
              other      146
              getz        49
mazda         3_reihe    164
              6_reihe    161
              other      138
mercedes_benz c_klasse  1136
              e_klasse   958
              a_klasse   539
nissan        micra      243
              other      121
              almera      92
opel          corsa     1592
              astra     1348
              vectra     544
peugeot       2_reihe    600
              3_reihe    322
              1_reihe    152
renault       twingo     615
              clio       473
              megane     335
seat          ibiza      328
              leon       194
              arosa      104
skoda         octavia    305
              fabia      280
              superb      51
smart         fortwo     550
              forfour     45
              roadster    32
toyota        yaris      126
              other      106
              corolla     89
volkswagen    golf      3707
              polo      1609
              passat    1349

## Conclusion

We have done cleaning in dataset and performed analysis by comparing features with each other. Out of 35 code cells 25 are dedicated to exploring and cleaning the data, so if it´s not 80% of all the time that a data scientist spends to clean and prepare the data, it´s 71%, still a very important part of job. 

In the analysis conducted we found out that:
 - Volkswagen is the most represented brand in our dataset together with it´s top model Volkswagen Golf. Probably the explication might be in an optimal relation between the price you can get and the total mileage of the car;
 - the cars with unrepaired damage are sold about 60% cheaper than the cars with no damage to repair;
 - putting your car on sale after riding 121000 might also drop the price by 50%.
 
There´re always some more steps to take:
 - his analysis only considers price and mileage as indicators of quality. There are potential other factors such as `registration_year` that this analysis did not consider. Future analysis could incorporate this variable.
 - we´ve analysed only about 46000 entries of the webpage, in order to get more precise conclusion it´s possible to conduct the analysis on the original sample of 370,000 to determine if the findings hold.